server/src/main/java/org/elasticsearch/action/support/RetryableAction.java

```diff
@@ -0,0 +1,136 @@
+        this.retryExecutor = retryExecutor;
+    }
+
+    public void run() {
+        runTryAction(new RetryingListener(initialDelayMillis, null));
+    }
+
+    public void cancel(Exception e) {
+        if (isDone.compareAndSet(false, true)) {
+            finalListener.onFailure(e);
+        }
+
+    }
+
+    private void runTryAction(ActionListener<Response> listener) {
+        try {
+            tryAction(listener);
+        } catch (Exception e) {
+            listener.onFailure(e);
+        }
+    }
+
+    public abstract void tryAction(ActionListener<Response> listener);
+
+    public abstract boolean shouldRetry(Exception e);
+
+    private class RetryingListener implements ActionListener<Response> {
+
+        private final long nextDelayMillis;
+        private final Exception existingException;
+
+        private RetryingListener(long nextDelayMillis, Exception existingException) {
+            this.nextDelayMillis = nextDelayMillis;
+            this.existingException = existingException;
+        }
+
+        @Override
+        public void onResponse(Response response) {
+            if (isDone.compareAndSet(false, true)) {
+                finalListener.onResponse(response);
+            }
+        }
+
+        @Override
+        public void onFailure(Exception e) {
+            if (shouldRetry(e)) {
+                final long elapsedMillis = threadPool.relativeTimeInMillis() - startMillis;
+                if (elapsedMillis > timeoutMillis) {
+                    logger.debug(() -> new ParameterizedMessage("retryable action timed out after {}",
+                        TimeValue.timeValueMillis(elapsedMillis)), e);
+                    addExisting(e);
+                    if (isDone.compareAndSet(false, true)) {
+                        finalListener.onFailure(e);
+                    }
+                } else {
+                    logger.debug(() -> new ParameterizedMessage("retrying action that failed in {}",
+                        TimeValue.timeValueMillis(nextDelayMillis)), e);
+                    addExisting(e);
+                    Runnable runnable = () -> runTryAction(new RetryingListener(nextDelayMillis * 2, e));
+                    final long midpoint = (nextDelayMillis / 2);
+                    final int randomness = Randomness.get().nextInt((int) Math.min(midpoint, Integer.MAX_VALUE));
+                    final long delayMillis = midpoint + randomness;
+                    if (isDone.get() == false) {
+                        threadPool.schedule(runnable, TimeValue.timeValueMillis(delayMillis), retryExecutor);
+                    }
+                }
+            } else {
+                addExisting(e);
+                if (isDone.compareAndSet(false,true)) {
+                    finalListener.onFailure(e);
+                }
+            }
+        }
+
+        private void addExisting(Exception e) {
+            if (existingException != null) {
+                e.addSuppressed(existingException);
```

Comment:
I think we need to limit the suppressed exceptions in case we retry too many times.